lte/gateway/c/oai/tasks/s1ap/s1ap_mme_nas_procedures.c

```diff
@@ -731,167 +851,204 @@ void s1ap_handle_conn_est_cnf(
+        conn_est_cnf_pP->ue_security_capabilities_integrity_algorithms);
+    ASN_SEQUENCE_ADD(&out->protocolIEs.list, ie);
   }
-
-  initialContextSetupRequest_p->ueSecurityCapabilities.encryptionAlgorithms
-      .buf = calloc(2, sizeof(uint8_t));
-  memcpy(
-      initialContextSetupRequest_p->ueSecurityCapabilities.encryptionAlgorithms
-          .buf,
-      (uint8_t*) &conn_est_cnf_pP
-          ->ue_security_capabilities_encryption_algorithms,
-      2);
-  initialContextSetupRequest_p->ueSecurityCapabilities.encryptionAlgorithms
-      .size = 2;
-  initialContextSetupRequest_p->ueSecurityCapabilities.encryptionAlgorithms
-      .bits_unused = 0;
-  initialContextSetupRequest_p->ueSecurityCapabilities
-      .integrityProtectionAlgorithms.buf = calloc(2, sizeof(uint8_t));
-  memcpy(
-      initialContextSetupRequest_p->ueSecurityCapabilities
-          .integrityProtectionAlgorithms.buf,
-      (uint8_t*) &conn_est_cnf_pP
-          ->ue_security_capabilities_integrity_algorithms,
-      2);
-  initialContextSetupRequest_p->ueSecurityCapabilities
-      .integrityProtectionAlgorithms.size = 2;
-  initialContextSetupRequest_p->ueSecurityCapabilities
-      .integrityProtectionAlgorithms.bits_unused = 0;
-  OAILOG_DEBUG_UE(
-      LOG_S1AP, imsi64, "security_capabilities_encryption_algorithms 0x%04X\n",
-      conn_est_cnf_pP->ue_security_capabilities_encryption_algorithms);
-  OAILOG_DEBUG_UE(
-      LOG_S1AP, imsi64, "security_capabilities_integrity_algorithms 0x%04X\n",
-      conn_est_cnf_pP->ue_security_capabilities_integrity_algorithms);
-
+  /* mandatory */
+  ie = (S1ap_InitialContextSetupRequestIEs_t*) calloc(
+      1, sizeof(S1ap_InitialContextSetupRequestIEs_t));
+  ie->id            = S1ap_ProtocolIE_ID_id_SecurityKey;
+  ie->criticality   = S1ap_Criticality_reject;
+  ie->value.present = S1ap_InitialContextSetupRequestIEs__value_PR_SecurityKey;
   if (conn_est_cnf_pP->kenb) {
-    initialContextSetupRequest_p->securityKey.buf =
-        calloc(AUTH_KENB_SIZE, sizeof(uint8_t));
+    ie->value.choice.SecurityKey.buf = calloc(AUTH_KENB_SIZE, sizeof(uint8_t));
     memcpy(
-        initialContextSetupRequest_p->securityKey.buf, conn_est_cnf_pP->kenb,
+        ie->value.choice.SecurityKey.buf, conn_est_cnf_pP->kenb,
         AUTH_KENB_SIZE);
-    initialContextSetupRequest_p->securityKey.size = AUTH_KENB_SIZE;
+    ie->value.choice.SecurityKey.size = AUTH_KENB_SIZE;
   } else {
     OAILOG_DEBUG(LOG_S1AP, "No kenb\n");
-    initialContextSetupRequest_p->securityKey.buf  = NULL;
-    initialContextSetupRequest_p->securityKey.size = 0;
+    ie->value.choice.SecurityKey.buf  = NULL;
+    ie->value.choice.SecurityKey.size = 0;
   }
+  ie->value.choice.SecurityKey.bits_unused = 0;
+  ASN_SEQUENCE_ADD(&out->protocolIEs.list, ie);
 
-  initialContextSetupRequest_p->securityKey.bits_unused = 0;
+  /* optional */
+  /*
+   * Only add capability information if it's not empty.
+   */
+  if (conn_est_cnf_pP->ue_radio_cap_length) {
```

Comment:
Can we use the length info form ue_radio_capability, as datatype of ue_radio_capability is bstring.

Reply:
Resolved